pkg/apis/garden/validation/validation.go

```diff
@@ -312,41 +312,44 @@ func validateAlicloudMachineTypeConstraints(machineTypes []garden.AlicloudMachin
 	return allErrs
 }
 
-func validateMachineImageNames(names []string, fldPath *field.Path) field.ErrorList {
+func validateMachineImages(machineImages []garden.MachineImage, fldPath *field.Path) field.ErrorList {
 	allErrs := field.ErrorList{}
 
-	if len(names) == 0 {
+	if len(machineImages) == 0 {
 		allErrs = append(allErrs, field.Required(fldPath, "must provide at least one machine image"))
-		return allErrs
 	}
 
-	imageNames := map[string]bool{}
-	for i, name := range names {
-		idxPath := fldPath.Index(i)
+	latestMachineImages, err := helper.DetermineLatestMachineImages(machineImages)
```

Comment:
One very big problem with this approach is that [Strategic Merge Patch](https://kubernetes.io/docs/tasks/run-application/update-api-object-kubectl-patch/#notes-on-the-strategic-merge-patch) won't work.

Reply:
The correct way to introduce such behavior in the API is to add a new field called `versions` and follow the [API compatibility guideline](https://github.com/kubernetes/community/blob/dfcdfd68206d1ca6a8fa02b55e6aef98b6fd0725/contributors/devel/sig-architecture/api_changes.md#on-compatibility).

Reply:
I guess we can live with that considering the fact that we will introduce new APIs for `Shoot`s, `CloudProfile`s, etc. soon anyways.

Reply:
There should be a conversion for those APIs and this change is going to make this conversion more error-prone and difficult. 

It's way better to introduce this field `versions` now and the conversion later one is going to be quite easy.

Reply:
I don't see a reason for doing this now as we anyways have to invest a lot of time to implement the new APIs. The PR is almost ready as it is, so investing this time now would block it from getting merged. Sure, it would simplify the conversion later because the code would already be there. If @danielfoehrKn want's to do it now then I'm fine with it, but as said, I don't see the need now.

Reply:
> I don't see a reason for doing this now as we anyways have to invest a lot of time to implement the new APIs. The PR is almost ready as it is, so investing this time now would block it from getting merged.

We've cut corners once with this image and our end-users had to fix things.
if we cut corner end-users have to fix lots more things when converting versions.

So **NO** cutting corners.

Reply:
I cannot judge the time effort for implementation atm as I am not clear how to implement it.
if you have time @mvladev we can discuss the details outside if this review.

Reply:
@mvladev Can you elaborate which exact problems you see when we don't do it now? Apart from not following the API conventions (yes, we should follow them, but considering the fact that we will have new APIs soon where we will do this I don't think this is a problem) the only thing you mentioned was that strategic merge patch is not possible. End-users cannot update `CloudProfile`s anyways, and I doubt that anybody setting up Gardener is relying on it. If somebody needs it urgently we can always give priority to introduce `versions` even in the current (old) API, but I wouldn't do it proactively. Again, my main motivation is the time effort that we anyways need to spend soon to get the new APIs, so it's not worth splitting it up and doing this little part now. Can we please go ahead with the actual feature of the PR?

Reply:
> yes, we should follow them, but considering the fact that we will have new APIs soon where we will do this I don't think this is a problem

Having new APIs doesn't mean that we should break existing ones. Changing the validation in such way changes the API behavior. The `v1beta1` API should be supported for quite a while - first the new `API` should be introduce in 1 release, then in the next be enabled by default, then in `xxx` releases the old API should be disabled (as NOT removed) and then finally after `xxx+1` `v1beta1` should be removed.

>  End-users cannot update CloudProfiles anyways

By `end-users` I meant API consumers - every user can interact with those APIs if given correct permissions.

Don't get me wrong - I like idea behind the proposed change, but let's do it properly which won't hurt API consumers and us when we introduce the new APIs

Reply:
> Having new APIs doesn't mean that we should break existing ones.

We don't break it, we simply don't follow the convention in a minor detail.

> Changing the validation in such way changes the API behavior.

Changing the API behaviour is not a problem as long as things that were possible before are still possible. This is the case. This PR simply allows *more* that was not allowed previously. This is no breaking change.

> The v1beta1 API should be supported for quite a while - first the new API should be introduce in 1 release, then in the next be enabled by default, then in xxx releases the old API should be disabled (as NOT removed) and then finally after xxx+1 v1beta1 should be removed.

That's fine and we'll do it like this, it's clear that many end-users depend on the current `garden.sapcloud.io/v1beta1` and will need some time to adapt to the new API.

> Don't get me wrong - I like idea behind the proposed change, but let's do it properly which won't hurt API consumers and us when we introduce the new APIs

If you insist that much then please contact @danielfoehrKn and help him doing the change. The issue https://github.com/gardener/gardener/issues/1117 (with a proposal describing exactly the currently implemented API behaviour) as well as the PR are open for quite some time. Please help get it done asap.

Reply:
Backwards compatibility has been implemented & is ready for review